src/libaktualizr/primary/aktualizr.cc

```diff
@@ -83,14 +83,32 @@ bool Aktualizr::UptaneCycle() {
 std::future<void> Aktualizr::RunForever() {
   std::future<void> future = std::async(std::launch::async, [&]() {
     SendDeviceData().get();
-    while (UptaneCycle()) {
-      std::this_thread::sleep_for(std::chrono::seconds(config_.uptane.polling_sec));
+
+    std::unique_lock<std::mutex> l(exit_cond_.m);
+    while (true) {
+      if (!UptaneCycle()) {
+        break;
+      }
+
+      if (exit_cond_.cv.wait_for(l, std::chrono::seconds(config_.uptane.polling_sec),
+                                 [this] { return exit_cond_.flag; })) {
+        break;
+      }
     }
     uptane_client_->completeInstall();
   });
   return future;
 }
 
+void Aktualizr::Shutdown() {
+  Abort();
```

Comment:
Abort() stops the command queue threads which is stopped in CommandQueue() dtor which itself is triggered during Aktualizr instance destruction, so calling Abort() in case of signal handling from the app/exe that destroys Aktualizr instance anyway looks like redundant. Having said that, I think it, might be useful if there is a need in calling RunForever() and Shutdown() multiple times within a single Aktualizr instance lifetime. Perhaps, break it into two methods ?

Reply:
`Abort()` docstring says that it aborts the currently running command if it can be aborted, that's why it sounded appropriate to call it from the signal handling.

So you're suggesting moving away `Abort()` from this method to the signal handler which will call `Abort();Shutdown();`?

Reply:
Or given the docstring I should maybe merge Shutdown() into Abort(): Abort() would also apply to RunForever()?

Reply:
I think, it depends on what Shutdown() is intended for ? If we would like to stop Aktualizr during an app exit then this method is useless at all because teardown/shutdown happens at Aktualizr dtor.

If Shutdown() is intended for the case when an user/app would like to stop aktualizr's UptaneCycle and then run it again within a single Aktualizr instance lifetime then the user will need to Initialize() if abort is part of Shutdown. BTW, abort doesn't stop the report queue thread.

Whether abort() should be part of the signal handler routine depends on a need in aborting ongoing operation during restart. I don't have strong opinion here, taking into account that the most timing operation is download and we support download resuming I think, I am fine with abort, but it might have negative impact when some write to DB or communication with secondaries is interrupted. I think, I would go with abort() at the moment.

Reply:
I'm not sure what you exactly mean in your first point. You mean that Shutdown() wouldn't be useful as an handler to a gui event for example? And you're saying that calling Abort() there is not useful because it's also called later? But the point is that calling Abort() early can speed up the tear down and trigger the `~Aktualizr()` call in a shorter delay. Or am I missing something

But your second point sounds right, calling `Initialize()` again after `Shutdown()` sounds quite wrong. I'll move `Abort()` to the signal handler.